### PR TITLE
Fix compiler Bug in local copy propagation #1985

### PR DESCRIPTION
**Pull request by milicaperic**
Change made in makeSideEffectStatement where the case which processed binary operator was missing.
After that a new bug appeared because isValid has not yet been processed in convertActionBody.

Change made:

In makeSideEffectStatement:
-Processing of unary operation was replaced with processing of unary and binary
-The given example illustrates the situation where the change needs to be applied

~~~ P4
control c3(inout headers hdr,inout metadata meta,inout standard_metadata_t std_meta) {
    action a(in bool b) {
        hdr.h.x = 8w0;
    }
    table t {
        key = { }
        actions = {
            a(hdr.h.isValid() || (bool) 1w1);
        }
    }
    apply {
        t.apply();
    }
}
~~~

or more complicated:

~~~ P4
control c3(inout headers hdr,inout metadata meta,inout standard_metadata_t std_meta) {
    action a(in bool b) {
        hdr.h.x = 8w0;
    }
    table t {
        key = { }
        actions = {
            a(!(!hdr.h.isValid() || !hdr.g.isValid()) ||  !(!hdr.f.isValid() || !hdr.e.isValid()) ||  !hdr.i.isValid());
        }
    }
    apply {
        t.apply();
    }
}
~~~

In convertActionBody:
-Added case that handles isValid
-Set prim with "no_op"
-The way we use isValid allows us to ignore the validity as opposed to the if-statement where this property is crucial
as in example:

    action a(in bool b) {
    if(b)
        hdr.h.x = 8w0;
    }
    table t {
        key = { }
        actions = {
            a(!(!hdr.h.isValid() || !hdr.g.isValid()) ||  !(!hdr.f.isValid() || !hdr.e.isValid()) ||  !hdr.i.isValid());
        }
    }
    apply {
        t.apply();
    }

    where no needed for makeSideEffectStatement because local variable is live

### CODE REVIEW

**Comment by jafingerhut** [comment]
Should the test program that fails with the latest p4c code, without this fix, from issue #1985 be added to this PR, with expected test outputs?  It seems a bad idea to leave out an existing known test program, when there is one.

**Comment by milicaperic** [comment]
> Should the test program that fails with the latest p4c code, without this fix, from issue #1985 be added to this PR, with expected test outputs? It seems a bad idea to leave out an existing known test program, when there is one.

You are right I added two tests for verifying solution

**Comment by mbudiu-vmw** [comment]
@ChrisDodd : do you want to take a look at this?

**Comment by ChrisDodd** [comment]
will do (was on vacation last week, so missed this).

File: midend/local_copyprop.cpp
Comment (ChrisDodd) [inline midend/local_copyprop.cpp:48]: This is incorrect if both `->left` and `->right` have side effects (the side effects in `->right` will be lost).  Add a `BUG_CHECK(!hasSideEffects(binary->right)`... at line 47 at least.

**Comment by ChrisDodd** [comment]
So my suggested "right" fix would be in `hasSideEffects::preorder(MethodCallExpression`, if the `mi` returned by `MethodInstance::resolve` is a `BuiltInMethod` with the name `"isValid"`, return true and don't set result.

**Comment by milicaperic** [comment]
Thank you for suggestion @ChrisDodd , I've made requested changes.

**Comment by milicaperic** [comment]
Is there anything more that I should change in order to merge this PR?

**Comment by mbudiu-vmw** [comment]
We are waiting for another review from @ChrisDodd

**Review by ChrisDodd — APPROVED** [review]

**Pull request merged**
